File: backend/monitor/openapi/accounts.py
"""OpenAPI schemas for accounts, dashboards, and account status."""

from __future__ import annotations

from .common import _nullable
from .cpa_status import cpa_status_schema


def account_schemas(nullable_number: dict, nullable_string: dict) -> dict:
    return {
        "MonitoredAccount": {
            "type": "object",
            "required": [
                "id",
                "provider",
                "source_account_id",
                "external_account_id",
                "cpa_auth_index",
                "pool_id",
                "name",
                "enabled",
                "quota_query_mode",
                "quota_profile",
                "detected_plan_type",
                "effective_quota_profile",
                "capacity_min_usd_override",
                "capacity_max_usd_override",
                "capacity_min_usd",
                "capacity_max_usd",
                "last_local_check_at",
                "last_upstream_check_at",
                "last_success_at",
                "next_local_check_at",
                "last_error",
            ],
            "properties": {
                "id": {"type": "integer"},
                "provider": {
                    "type": "string",
                    "enum": ["sub2api", "cpa"],
                },
                "source_account_id": {"type": "string"},
                "external_account_id": _nullable("integer"),
                "cpa_auth_index": nullable_string,
                "pool_id": {"type": "integer"},
                "name": {"type": "string"},
                "enabled": {"type": "boolean"},
                "quota_query_mode": {
                    "type": "string",
                    "enum": ["passive", "direct"],
                },
                "quota_profile": {
                    "type": "string",
                    "enum": ["auto", "plus", "pro_5x", "pro_20x"],
                },
                "detected_plan_type": {
                    "type": "string",
                    "enum": ["", "plus", "pro"],
                },
                "effective_quota_profile": {
                    "type": "string",
                    "enum": ["plus", "pro_5x", "pro_20x"],
                },
                "capacity_min_usd_override": nullable_number,
                "capacity_max_usd_override": nullable_number,
                "capacity_min_usd": {"type": "number"},
                "capacity_max_usd": {"type": "number"},
                "last_local_check_at": nullable_string,
                "last_upstream_check_at": nullable_string,
                "last_success_at": nullable_string,
                "next_local_check_at": nullable_string,
                "last_error": {"type": "string"},
            },
        },
        "Dashboard": {
            "type": "object",
            "required": [
                "configured",
                "monitoring_enabled",
                "accounts",
                "selected_account_id",
                "selected_provider",
                "last_local_check_at",
                "last_upstream_check_at",
                "snapshot_stale",
                "last_success_at",
                "last_error",
                "quota_query_mode",
                "sub2api_admin_url",
                "upstream_admin_url",
                "fast_correction_enabled",
                "weekly_quota_model",
                "cycle",
                "participants",
                "needs_manual_update_count",
            ],
            "properties": {
                "cpa_summary": {"oneOf": [{"$ref": "#/components/schemas/CPAPoolSummary"}, {"type": "null"}]},
                "configured": {"type": "boolean"},
                "monitoring_enabled": {"type": "boolean"},
                "accounts": {
                    "type": "array",
                    "items": {"$ref": "#/components/schemas/MonitoredAccount"},
                },
                "selected_account_id": _nullable("integer"),
                "selected_provider": {
                    "type": ["string", "null"],
                    "enum": ["sub2api", "cpa", None],
                },
                "last_local_check_at": nullable_string,
                "last_upstream_check_at": nullable_string,
                "snapshot_stale": {"type": "boolean"},
                "last_success_at": nullable_string,
                "last_error": {"type": "string"},
                "quota_query_mode": {
                    "type": ["string", "null"],
                    "enum": ["passive", "direct", None],
                },
                "sub2api_admin_url": {"type": "string", "format": "uri"},
                "upstream_admin_url": {"type": "string", "format": "uri"},
                "fast_correction_enabled": {"type": "boolean"},
                "weekly_quota_model": {
                    "type": "string",
                    "enum": ["time_varying", "constant_average"],
                },
                "cycle": {
                    "oneOf": [
                        {"$ref": "#/components/schemas/DashboardCycle"},
                        {"type": "null"},
                    ]
                },
                "participants": {
                    "type": "array",
                    "items": {"$ref": "#/components/schemas/Participant"},
                },
                "needs_manual_update_count": {"type": "integer"},
            },
        },
        "DashboardCycle": {
            "type": "object",
            "required": [
                "id",
                "observed_at",
                "starts_at",
                "resets_at",
                "upstream_used_percent",
                "interval_used_percent",
                "effective_usd_per_percent",
                "selected_total_cost",
                "selected_total_cost_breakdown",
                "start_cost_breakdown",
                "unattributed_used_percent",
                "sample_note",
                "snapshot_sampled_at",
                "rate_calculated",
                "estimated_used_percent",
                "capacity_lower_usd",
                "capacity_upper_usd",
                "model_diagnostics",
            ],
            "properties": {
                "id": {"type": "integer"},
                "observed_at": {"type": "string", "format": "date-time"},
                "starts_at": nullable_string,
                "resets_at": {"type": "string", "format": "date-time"},
                "upstream_used_percent": {"type": "number"},
                "interval_used_percent": {"type": "number"},
                "effective_usd_per_percent": nullable_number,
                "selected_total_cost": {"type": "number"},
                "selected_total_cost_breakdown": {
                    "$ref": "#/components/schemas/CostBreakdown"
                },
                "start_cost_breakdown": {"$ref": "#/components/schemas/CostBreakdown"},
                "unattributed_used_percent": {"type": "number"},
                "sample_note": {"type": "string"},
                "snapshot_sampled_at": nullable_string,
                "rate_calculated": {"type": "boolean"},
                "estimated_used_percent": {"type": "number"},
                "capacity_lower_usd": nullable_number,
                "capacity_upper_usd": nullable_number,
                "model_diagnostics": {"type": "object"},
            },
        },
        "CostBreakdown": {
            "type": "object",
            "required": [
                "sub2api_cost_usd",
                "fast_correction_usd",
                "total_cost_usd",
            ],
            "properties": {
                "sub2api_cost_usd": {"type": "number"},
                "fast_correction_usd": {"type": "number"},
                "total_cost_usd": {"type": "number"},
            },
        },
        "AccountStatus": {
            "type": "object",
            "required": [
                "configured",
                "sampled_at",
                "stats_days",
                "connection_error",
                "accounts",
            ],
            "properties": {
                "configured": {"type": "boolean"},
                "sampled_at": {"type": "string", "format": "date-time"},
                "stats_days": {"type": "integer"},
                "connection_error": nullable_string,
                "accounts": {
                    "type": "array",
                    "items": {"$ref": "#/components/schemas/AccountStatusAccount"},
                },
            },
        },
        "AccountStatusAccount": {
            "type": "object",
            "required": [
                "id",
                "provider",
                "source_account_id",
                "external_account_id",
                "name",
                "enabled",
                "quota_query_mode",
                "cycles",
                "runtime",
                "usage",
                "stats",
                "warnings",
                "temporary_disables",
            ],
            "properties": {
                "cpa_quota": cpa_status_schema(),
                "id": {"type": "integer"},
                "provider": {
                    "type": "string",
                    "enum": ["sub2api", "cpa"],
                },
                "source_account_id": {"type": "string"},
                "external_account_id": _nullable("integer"),
                "name": {"type": "string"},
                "enabled": {"type": "boolean"},
                "quota_query_mode": {
                    "type": "string",
                    "enum": ["passive", "direct"],
                },
                "cycles": {
                    "type": "array",
                    "items": {
                        "$ref": "#/components/schemas/AccountCycleUsage"
                    },
                },
                "runtime": {
                    "oneOf": [
                        {"$ref": "#/components/schemas/AccountRuntimeStatus"},
                        {"type": "null"},
                    ]
                },
                "usage": {
                    "oneOf": [
                        {"$ref": "#/components/schemas/AccountUsageStatus"},
                        {"type": "null"},
                    ]
                },
                "stats": {
                    "oneOf": [
                        {"$ref": "#/components/schemas/AccountUsageStats"},
                        {"type": "null"},
                    ]
                },
                "warnings": {
                    "type": "array",
                    "items": {"type": "string"},
                },
                "temporary_disables": {
                    "type": "array",
                    "items": {
                        "$ref": "#/components/schemas/TemporaryDisable"
                    },
                },
            },
        },
        "TemporaryDisable": {
            "type": "object",
            "required": [
                "id",
                "account_id",
                "scope",
                "model",
                "started_at",
                "restore_at",
                "retry_at",
                "restored_at",
                "restore_source",
                "created_by",
                "last_error",
            ],
            "properties": {
                "id": {"type": "integer"},
                "account_id": {"type": "integer"},
                "scope": {
                    "type": "string",
                    "enum": ["account", "model"],
                    "description": "account 表示整个账号暂停调度，model 表示从模型白名单移除一个模型。",
                },
                "model": {
                    "type": "string",
                    "description": "被禁用的模型名；账号级禁用为空字符串。",
                },
                "started_at": {"type": "string", "format": "date-time"},
                "restore_at": {
                    "type": "string",
                    "format": "date-time",
                    "description": "计划恢复时刻；后台到点后写回上游。",
                },
                "retry_at": {
                    "oneOf": [
                        {"type": "string", "format": "date-time"},
                        {"type": "null"},
                    ],
                    "description": "自动恢复失败后的下次尝试时刻。",
                },
                "restored_at": _nullable("string"),
                "restore_source": {
                    "type": "string",
                    "enum": ["", "manual", "auto"],
                },
                "created_by": {
                    "type": "string",
                    "description": "发起禁用的管理员；只读接口与系统用户始终为空字符串。",
                },
                "last_error": {
                    "type": "string",
                    "description": "未经确认的上游写入；为空表示上游已确认。",
                },
            },
        },
        "AccountCycleUsage": {
            "type": "object",
            "required": [
                "sequence",
                "started_at",
                "ended_at",
                "used_percent",
                "used_usd",
                "is_current",
            ],
            "properties": {
                "sequence": {"type": "integer"},
                "started_at": {"type": "string", "format": "date-time"},
                "ended_at": {"type": "string", "format": "date-time"},
                "used_percent": {"type": "number"},
                "used_usd": {"type": "number"},
                "is_current": {"type": "boolean"},
            },
        },
        "AccountRuntimeStatus": {
            "type": "object",
            "properties": {
                "name": nullable_string,
                "account_type": nullable_string,
                "status": nullable_string,
                "schedulable": _nullable("boolean"),
                "current_concurrency": _nullable("integer"),
                "concurrency_limit": _nullable("integer"),
                "last_used_at": nullable_string,
                "rate_limited_at": nullable_string,
                "rate_limit_reset_at": nullable_string,
                "overload_until": nullable_string,
                "temp_unschedulable_until": nullable_string,
                "temp_unschedulable_reason": nullable_string,
                "error_message": nullable_string,
            },
        },
        "AccountUsageWindow": {
            "type": "object",
            "properties": {
                "used_percent": nullable_number,
                "reset_at": nullable_string,
                "remaining_seconds": _nullable("integer"),
                "request_count": _nullable("integer"),
                "token_count": _nullable("integer"),
                "account_cost_usd": nullable_number,
                "standard_cost_usd": nullable_number,
                "user_cost_usd": nullable_number,
            },
        },
        "AccountUsageStatus": {
            "type": "object",
            "properties": {
                "source": nullable_string,
                "updated_at": nullable_string,
                "five_hour": {
                    "oneOf": [
                        {"$ref": "#/components/schemas/AccountUsageWindow"},
                        {"type": "null"},
                    ]
                },
                "seven_day": {
                    "oneOf": [
                        {"$ref": "#/components/schemas/AccountUsageWindow"},
                        {"type": "null"},
                    ]
                },
                "needs_verify": _nullable("boolean"),
                "is_banned": _nullable("boolean"),
                "needs_reauth": _nullable("boolean"),
                "error_code": nullable_string,
                "error": nullable_string,
            },
        },
        "AccountUsageStats": {
            "type": "object",
            "properties": {
                "account_cost_usd": nullable_number,
                "fast_correction_usd": nullable_number,
                "account_cost_with_fast_correction_usd": nullable_number,
                "standard_cost_usd": nullable_number,
                "user_cost_usd": nullable_number,
                "request_count": _nullable("integer"),
                "token_count": _nullable("integer"),
                "avg_daily_cost_usd": nullable_number,
                "avg_daily_request_count": nullable_number,
                "avg_daily_token_count": nullable_number,
                "avg_duration_ms": nullable_number,
                "today": {
                    "type": ["object", "null"],
                    "additionalProperties": True,
                },
            },
        },
        "MonitoredAccountSummary": {
            "type": "object",
            "required": [
                "id",
                "provider",
                "source_account_id",
                "external_account_id",
                "name",
            ],
            "properties": {
                "id": {"type": "integer"},
                "provider": {
                    "type": "string",
                    "enum": ["sub2api", "cpa"],
                },
                "source_account_id": {"type": "string"},
                "external_account_id": _nullable("integer"),
                "name": {"type": "string"},
            },
        },
    }

File: frontend/src/types/accounts.ts
import type { CPACapacityEstimate } from "./cpa";
import type { CorrectionBreakdown } from "./common";
export type QuotaProfile = "auto" | "plus" | "pro_5x" | "pro_20x";
export type EffectiveQuotaProfile = Exclude<QuotaProfile, "auto">;
export type AccountProvider = "sub2api" | "cpa" | "gpt_load";

export interface MonitoredAccount {
  id: number;
  provider: AccountProvider;
  source_account_id: string;
  pool_id: number;
  external_account_id: number | null;
  cpa_auth_index: string | null;
  gpt_load_group_id?: number | null;
  gpt_load_credential_id?: number | null;
  gpt_load_cutover_at?: string | null;
  gpt_load_logs_synced_through?: string | null;
  name: string;
  enabled: boolean;
  quota_query_mode: "passive" | "direct";
  quota_profile: QuotaProfile;
  detected_plan_type: "" | "plus" | "pro";
  effective_quota_profile: EffectiveQuotaProfile;
  capacity_min_usd_override: number | null;
  capacity_max_usd_override: number | null;
  capacity_min_usd: number;
  capacity_max_usd: number;
  last_local_check_at: string | null;
  last_upstream_check_at: string | null;
  last_success_at: string | null;
  next_local_check_at: string | null;
  last_error: string;
}
export interface CPAAccountOption {
  auth_index: string;
  name: string;
  email: string;
  chatgpt_account_id: string;
  plan_type: string;
  status: string;
  status_message: string;
  disabled: boolean;
  unavailable: boolean;
  success: number;
  failed: number;
}
export interface GPTLoadAccountOption {
  group_id: number;
  group_name: string;
  channel_id: string;
  credential_id: number;
  email: string;
  mask: string;
  plan_type: string;
  configured_status: string;
  effective_status: string;
}
export interface AccountRuntimeStatus {
  name: string | null;
  account_type: string | null;
  status: string | null;
  schedulable: boolean | null;
  current_concurrency: number | null;
  concurrency_limit: number | null;
  last_used_at: string | null;
  rate_limited_at: string | null;
  rate_limit_reset_at: string | null;
  overload_until: string | null;
  temp_unschedulable_until: string | null;
  temp_unschedulable_reason: string | null;
  error_message: string | null;
}
export interface AccountUsageWindow {
  used_percent: number | null;
  reset_at: string | null;
  remaining_seconds: number | null;
  request_count: number | null;
  token_count: number | null;
  account_cost_usd: number | null;
  standard_cost_usd: number | null;
  user_cost_usd: number | null;
}
export interface AccountUsageStatus {
  source: string | null;
  updated_at: string | null;
  five_hour: AccountUsageWindow | null;
  seven_day: AccountUsageWindow | null;
  needs_verify: boolean | null;
  is_banned: boolean | null;
  needs_reauth: boolean | null;
  error_code: string | null;
  error: string | null;
}
export interface AccountUsageStats extends CorrectionBreakdown {
  account_cost_with_correction_usd?: number | null;
  correction_collected_until?: string | null;
  days: number | null;
  actual_days_used: number | null;
  account_cost_usd: number | null;
  fast_correction_usd: number | null;
  account_cost_with_fast_correction_usd: number | null;
  standard_cost_usd: number | null;
  user_cost_usd: number | null;

  request_count: number | null;
  token_count: number | null;
  avg_daily_cost_usd: number | null;
  avg_daily_request_count: number | null;
  avg_daily_token_count: number | null;
  avg_duration_ms: number | null;
  today: {
    date: string | null;
    account_cost_usd: number | null;
    user_cost_usd: number | null;
    request_count: number | null;
    token_count: number | null;
  } | null;
}
export interface AccountCycleUsage {
  sequence: number;
  started_at: string;
  ended_at: string;
  used_percent: number;
  used_usd: number;
  is_current: boolean;
}
export type TemporaryDisableScope = "account" | "model";
export interface TemporaryDisable {
  id: number;
  account_id: number;
  scope: TemporaryDisableScope;
  model: string;
  started_at: string | null;
  restore_at: string | null;
  retry_at: string | null;
  restored_at: string | null;
  restore_source: "" | "manual" | "auto";
  created_by: string;
  last_error: string;
}
export interface AccountStatusAccount {
  cpa_quota?: CPAQuotaDetail;
  id: number;
  provider: AccountProvider;
  source_account_id: string;
  external_account_id: number | null;
  name: string;
  enabled: boolean;
  quota_query_mode: "passive" | "direct";
  cycles: AccountCycleUsage[];
  runtime: AccountRuntimeStatus | null;
  usage: AccountUsageStatus | null;
  stats: AccountUsageStats | null;
  warnings: string[];
  temporary_disables: TemporaryDisable[];
}
export interface AccountStatusData {
  configured: boolean;
  sampled_at: string;
  stats_days: number;
  connection_error: string | null;
  accounts: AccountStatusAccount[];
}
export interface OpenAIAccountOption {
  id: number;
  name: string;
  type: string;
  status: string;
  schedulable: boolean;
}

export interface CPAQuotaMetrics {
  request_count: number;
  token_count: number;
  usage_usd: number;
  success_rate?: number | null;
  unpriced_request_count?: number;
}
export interface CPAQuotaPeriod {
  started_at: string;
  ended_at: string;
  metrics: CPAQuotaMetrics;
  coverage_complete: boolean;
  notice?: string;
}
export interface CPAQuotaDetail {
  totals: CPAQuotaPeriod;
  windows: {
    id: string;
    label: string;
    used_percent: number | null;
    reset_at: string | null;
    updated_at: string | null;
    source: string;
    boundary: "unknown" | "provider";
    current: CPAQuotaPeriod | null;
    previous: CPAQuotaPeriod | null;
    prediction: CPAQuotaMetrics | null;
    capacity_estimate?: CPACapacityEstimate | null;
    notice: string | null;
  }[];
  reset: {
    available_count: number | null;
    can_reset: boolean;
    history: {
      id: string;
      status: string;
      created_at: string;
      finished_at: string | null;
    }[];
  };
}
export interface CPAResetPreview {
  id: string;
  account_id: number;
  account_name: string;
  available_count: number;
  status: string;
  expires_at: string;
}
